### ISSUE TITLE
Docker health check

### ISSUE THREAD

**Issue by JerboaGobi** [unpoller/unpoller#406]
Just wondering if there were any plans to add any docker health checks? Or if you could suggest a health check that could be implemented via docker-compose?

**Comment by davidnewhall** [comment]
Have you been in a situation where a health check would have restarted the container to correct resolve an issue? Have not heard of this yet.

**Comment by Taverius** [comment]
@davidnewhall at least for me, a healthcheck is not for having the container restart.

I mean, I might set that up in a pinch, but if you have to there's usually other problems with other solutions.

The reasons to have a check built in are:

1. While its easy enough to bind-mount `busybox` as `sh` and `wget` in Docker Standalone, this gets much trickier when orchestrating.
2. Container dependency - like `depends_on:` in compose - help avoid some subtle annoying sporadic issues and they're more reliable if you can get the dependant to wait for the things it depends on to be properly started.
3. While its certainly possible to parse the container-specific metrics and/or logs for specific things going wrong and have prometheus/mimir/grafana alertmanager or whatever send you a message, and you should do that in a production environment, what's even better and more reliable and way less likely to break in unexpected ways is to check every interval for `unhealthy` containers and send an alert.
4. Custom builds are certainly possible, but they add points of failure.

Basically, I think they're best used for more reliable startup sequencing and as a very basic level, highly reliable "Something's jacked, you should look at the logs" blinkenlight.

In the context of `unpoller`, statically compiled `curl` and some kind of shell for the proper `|| exit 1` , or `busybox`, are just way too big. The unpoller image is what, 18MB? Static Musl `curl` is like 8MB.

An option on the main executable would be the smallest and easiest to maintain - i.e. you just use:
```yaml
healthcheck:
  test: [ "CMD", "/usr/bin/unpoller", "ready" ]
```
and it exits 0|1 as appropriate, no shell needed, no extra dependencies to maintain, etc.

**Comment by davidnewhall** [comment]
Making the binary capable of checking and returning the web server health seems reasonable.

I would caution against using `depends` on unpoller. It has no startup dependencies, and if you're making other containers depend on it, that's probably not accurate.  Prometheus and Influx, for instance, do not depend on unpoller for startup either.

For what it's worth, I use Notifiarr for alerts when my containers go down.

**Comment by davidnewhall** [comment]
I've never been a real fan of adding stuff to the image. It's a single binary and you can re-package it into another container with a pretty small Dockerfile if you need the extras. Most folks don't and I like to keep it simple.

**Comment by Taverius** [comment]
Absolutely, for all the stated above reasons - size, attack surface, etc - but also specifically for HTTP API reading metrics infrastructure images, of which there are many:

* Why include an external app to perform an HTTP GET and see if the return code is correct for an application *which is itself already regularly doing so as part of its normal operations?*

Its a little bit baffling and yet I see plenty of projects either provide multiple types of ready endpoints but no way to query them internally - which adds an external dependency to check which adds a point of failure, and this in k8s-focused projects which are already complex enough to deploy - or they just blindly bundle `wget`.

I've seen the details of the internal check discussed a few times - do you just run the check internally, or do you actually use your base code to do an actual honest HTTP GET on the ready endpoint you're serving.

Personally, my point of view is "whichever leads to the least amount of custom code for the alive check"; imo minimizing the chance the check itself will break, and thus maintenance, trumps other concerns.

**Comment by platinummonkey** [comment]
I’ll look at this but what would it check? Just that it’s running and started? That *any* site is getting data or *all* sites? What about outputs?

**Comment by davidnewhall** [comment]
I was envisioning three endpoints. One to check only the web server, one that checks backend connectivity, and one that checks controller connectivity. The end user can choose what they want.

**Comment by Taverius** [comment]
The basic one is "is it live and the config checked out ok and it's not sitting there spitting out errors onto log because of some basic-tier misconfiguration from deployment/upgrading".

This lets you do basic dependency resolution, and is usually the one polled in container-self-health-checks.

Anything else is best-effort, and usually done via separate http endpoints like `/ready` so you can use it in shell scripts in other systems/containers, though appreciated.

**Comment by platinummonkey** [comment]
added in #890 will cut a release later
